### PR TITLE
Added register, signin, signout, and authentication for users

### PR DESCRIPTION
**Pull request by cuongla**
I have added register, signin (with JWT), signout, and authentication middleware for user.

### CODE REVIEW

**Review by JamesClark123 — APPROVED** [review]

File: .gitignore [hunk 1]
```diff
@@ -18,6 +18,8 @@ npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
 
+#keys
+dev.js
```
Comment (skassam21) [inline .gitignore:22]: what's in this file? seems like a strange file to gitignore

File: server/controllers/auth.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+exports.logout = async (req, res) => {
+  try {
+      req.user.tokens = req.user.tokens.filter(token => {
+          return token.token !== req.token;
+      });
+
+      await req.user.save();
+      res.send();
+  } catch(e) {
+      res.status(500).send();
```
Comment (skassam21) [inline server/controllers/auth.js]: I think Aecio commented about this in another PR

**Review by skassam21 — COMMENTED** [review]
because this was not merged earlier... there is a lot of overlap with existing PR - I guess you should close all the old PRs

File: .gitignore (hunk 1 above)
Comment (cuongla) [reply .gitignore:22]: This one is where I hold all aws keys, mongodb and secret. I didn't .env in this project to store keys

File: server/routes/user.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+);
+
+// @route PUT users/unfollow
+router.put(
+    '/me/unfollow', 
+    authMiddleware, 
+    UserController.removeFollowing,
+    UserController.removeFollower
+)
+module.exports = router;
```
Comment (skassam21) [inline server/routes/user.js:36]: there is a lot of space undernearth here

File: server/routes/user.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+router.put(
+    '/me/follow', 
+    authMiddleware, 
+    UserController.addFollowing,
+    UserController.addFollower
+);
+
+// @route PUT users/unfollow
+router.put(
+    '/me/unfollow',
```
Comment (skassam21) [inline server/routes/user.js:31]: This is not a good naming convention in this file (all the routes) - try to use a RESTful naming convention through here like how you did with the other routes, something like /users/ etc.

**Review by skassam21 — COMMENTED** [review]
Just address these issues and you can merge if things are working - this is a very large pull requests and ideally a lot of these functions would be in separate pull requests. You could work on separate tickets in separate pull requests

**Pull request closed without merging**